service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpConnectionHandler.java

```diff
@@ -0,0 +1,345 @@
+            {
+                throw new IOException("Timed out waiting for amqp links to open");
+            }
+
+            Thread.sleep(100);
+        }
+
+        if (protonJExceptionParser != null && protonJExceptionParser.getError() != null)
+        {
+            throw new IOException(String.format("Encountered an exception while opening connection: %s with description %s", protonJExceptionParser.getError(), protonJExceptionParser.getErrorDescription()));
+        }
+    }
+
+    private void openAsync() throws IOException
+    {
+        this.reactor = Proton.reactor(this);
+        new Thread(new ReactorRunner(this.reactor)).start();
+    }
+
+    public void close()
+    {
+        if (this.reactor != null)
+        {
+            log.debug("Closing amqp connection...");
+            this.reactor.stop();
+            this.reactor.free();
+            this.reactor = null;
+        }
+    }
+
+    @Override
+    public void onReactorInit(Event event)
+    {
+        event.getReactor().connection(this);
+    }
+
+    /**
+     * Event handler for the on delivery event
+     * @param event The proton event object
+     */
+    @Override
+    public void onDelivery(Event event)
+    {
+        if (event.getLink() instanceof Receiver)
+        {
+            log.trace("Received delivery on receiver link");
+            Receiver recv = (Receiver)event.getLink();
+            Delivery delivery = recv.current();
+
+            if (delivery.isReadable() && !delivery.isPartial() && delivery.getLink().getName().equalsIgnoreCase(tag))
+            {
+                int size = delivery.pending();
+                byte[] buffer = new byte[size];
+                int read = recv.recv(buffer, 0, buffer.length);
+                recv.advance();
+
+                org.apache.qpid.proton.message.Message msg = Proton.message();
+                msg.decode(buffer, 0, read);
+
+                DeliveryOutcome outcome = onMessageArrived(msg);
+
+                if (outcome == DeliveryOutcome.Complete)
+                {
+                    log.trace("Amqp message was completed, acknowledging it with Accepted");
+                    delivery.disposition(Accepted.getInstance());
+                    delivery.settle();
+                }
+                else if (outcome == DeliveryOutcome.Abandon)
+                {
+                    log.trace("Amqp message was abandoned, acknowledging it with Released");
+                    delivery.disposition(Released.getInstance());
+                    delivery.settle();
+                }
+                else if (outcome == DeliveryOutcome.Reject)
+                {
+                    //purposefully do nothing. Service will treat unack'd messages as rejected
+                    log.trace("Amqp message was rejected, will not acknowledge it");
+                }
+                else
+                {
+                    log.error("Unknown DeliveryOutcome provided, message will not be acknowledged");
```

Comment:
Switch-Case on the enum for better readability?

Reply:
Sure thing